Add a custom LLVM pass to replace fastmath multiple and add with muladd

yuyichao opened the pull request:
Now that it's our norm to use custom LLVM passes....

This currently does the replacement as long as one of the instruction allows unsafe arithmetic
mainly because LLVM vectorization pass does not always preserve the fastmath flags. This is arguable though I don't think there'll be any problem since this optimization is only done when the intermediate result is not being used or in another word when the `fmul` and `fadd` are producing only one result while allowing some contraction in this process....

I guess this kind of makes `muladd` obsolete, oh, well ;-p...

Fix #18654, Fix #22217

For simple cases:

```julia
julia> f1(a, b, c) = @fastmath a * b + c
f1 (generic function with 1 method)

julia> f2(a, b, c) = @fastmath a * b - c
f2 (generic function with 1 method)

julia> f3(a, b, c) = @fastmath c - a * b
f3 (generic function with 1 method)

julia> @code_native f1(1.0, 2.0, 3.0)
        .text
Filename: REPL[1]
        pushq   %rbp
        movq    %rsp, %rbp
        vfmadd213sd     %xmm2, %xmm1, %xmm0
Source line: 1
        popq    %rbp
        retq
        nopl    (%rax,%rax)

julia> @code_native f2(1.0, 2.0, 3.0)
        .text
Filename: REPL[2]
        pushq   %rbp
        movq    %rsp, %rbp
        vfmsub213sd     %xmm2, %xmm1, %xmm0
Source line: 1
        popq    %rbp
        retq
        nopl    (%rax,%rax)

julia> @code_native f3(1.0, 2.0, 3.0)
        .text
Filename: REPL[3]
        pushq   %rbp
        movq    %rsp, %rbp
        vfnmadd213sd    %xmm2, %xmm1, %xmm0
Source line: 1
        popq    %rbp
        retq
        nopl    (%rax,%rax)
```

For the complex case enabled by running this after the loop vectorizer
```julia
julia> function f(a, b)
           s = zero(eltype(a))
           @inbounds @simd for i in 1:length(a)
               s += a[i] * b[i]
           end
           return s
       end
f (generic function with 1 method)

julia> @code_llvm f(Float64[], Float64[])
...
  %34 = getelementptr double, double* %28, i64 12
  %35 = bitcast double* %34 to <4 x double>*
  %wide.load106 = load <4 x double>, <4 x double>* %35, align 8
  %36 = call <4 x double> @llvm.fmuladd.v4f64(<4 x double> %wide.load, <4 x double> %wide.load103, <4 x double> %vec.phi)
  %37 = call <4 x double> @llvm.fmuladd.v4f64(<4 x double> %wide.load100, <4 x double> %wide.load104, <4 x double> %vec.phi94)
  %38 = call <4 x double> @llvm.fmuladd.v4f64(<4 x double> %wide.load101, <4 x double> %wide.load105, <4 x double> %vec.phi95)
  %39 = call <4 x double> @llvm.fmuladd.v4f64(<4 x double> %wide.load102, <4 x double> %wide.load106, <4 x double> %vec.phi96)
  %index.next = add i64 %index, 16
  %40 = icmp eq i64 %index.next, %n.vec
  br i1 %40, label %middle.block, label %vector.body
...
```

yuyichao committed:
Add a custom LLVM pass to allow contraction of fast fadd/fsub with a normal fmul.

Fix #18654, Fix #22217

yuyichao commented:
@nanosoldier `runbenchmarks(ALL, vs = ":master")`

yuyichao commented on src/llvm-muladd.cpp:
```diff
@@ -0,0 +1,107 @@
+// This file is a part of Julia. License is MIT: https://julialang.org/license
+
+#define DEBUG_TYPE "lower_handlers"
```
Ooops. ;-p

vtjnash commented:
Worth submitting upstream? Of course, now that we're adding llvm passes, we're also adding llvm IR tests, which this will need one of :P

nanosoldier commented:
[Your benchmark job](https://github.com/JuliaLang/julia/pull/22262#issuecomment-306688797) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/8168f8fc162ae2a8daee699e2d5c1aa56e4d12b1/97f092e_vs_285ec1f/report.md). cc @jrevels

yuyichao commented:
> Worth submitting upstream?

Maybe after https://bugs.llvm.org/show_bug.cgi?id=33338 is fixed. It's unclear to me whether they'll want to have this in two places though....

> this will need one of

= = ...... I'll try. Adding test for LLVM PR's is hard enough......

yuyichao commented:
Test added...

Keno commented:
Is this actually beneficial? Why not turn on FMA fusion and rely on LLVM to do this at the MIR level?

yuyichao commented:
Can LLVM do that on per-instruction level and not per-function level?

Keno commented:
As of LLVM 5.0, yes. There's a new `contract` fastmath flag.

yuyichao commented:
That flag is what I'm using on LLVM 5.0 though I don't think having this work requires that? Anyway, it does seem like LLVM 5.0 can do that now so this won't be needed as much.

There's still a small difference. LLVM requires both the multiply and add to be contractable to merge the two so it doesn't create fma for the second case above.

yuyichao commented:
So it seems that LLVM will not contract normal `fmul` and fast `fadd` so I updated the pass to only do that (by marking the fmul(s) as `contract` on 5.0). I've also changed `muladd` to emit `contract` able fmul and fadd instead of `muladd` on 5.0 since it's easier for llvm to optimize. This may cause `muladd` to return slightly different result depending on the caller. Is that OK?

StefanKarpinski commented:
> This may cause `muladd` to return slightly different result depending on the caller.

That sounds upsetting. Can you elaborate a bit on when and how that would happen?

yuyichao commented:
`f(a, b, c) = (a * b, muladd(a, b, c))` can be optimized to `f(a, b, c) = (x = a * b; (x, x + c))`

StefanKarpinski commented:
But `f(a, b, c)` will not change behavior depending on where it appears? It's just `muladd` that can have different behaviors, right?

yuyichao commented:
> But `f(a, b, c)` will not change behavior depending on where it appears?

If it is inlined then it can (e.g. if the first result is not used). Essentially the semantic of `muladd` is changed to allow it be `fma` or not depend on which one the compiler think is faster.

StefanKarpinski commented:
This feels a bit dirty. But yet, if someone has explicitly used `muladd` that is essentially saying, "it doesn't matter if you do this operation as an `fma` or a multiply and an `add`"; we've interpreted this to apply to different systems, but I'm not sure why – if that's true – it wouldn't also be true in the same program. It either matters or it doesn't.

yuyichao commented:
So any more comment on this? I should note that inlining affecting result will also be true for `@fastmath` on LLVM 5.0 without this change and this is basically a more restricted version of it. In principle we can do this with a `@contractmath` macro though I'd rather not go that route (with its own complexity) unless there's good usecase for it that is not covered by `@fastmath`. (A contract ieee mode is certainly doable and should be easy though).

StefanKarpinski commented:
I'm ok with `@fastmath` doing this.

yuyichao commented:
So `muladd` is basically a more well defined `@fastmath`.

StefanKarpinski commented:
Sure, it's just a very specific instance of `@fastmath`.

musm commented:
Would it be possible to also  have this transformation in a more specific macro that just applies these transformations without including all the additional fastmath replacements? I think that might be useful.

yuyichao commented:
https://github.com/JuliaLang/julia/pull/22262#issuecomment-309908689

yuyichao commented:
Another example showing that this really isn't anything new and is actually already part of the `muladd` semantics from LLVM.

```julia
julia> muladd(1 - eps(Float64), 1 + eps(Float64), -1.0)
0.0

julia> f() = muladd(1 - eps(Float64), 1 + eps(Float64), -1.0)
f (generic function with 1 method)

julia> f()
-4.930380657631324e-32
```

yuyichao commented:
Given https://github.com/JuliaLang/julia/pull/22262#issuecomment-312272298 I just rebased and updated the doc clarifying that the result can be different depending on optimizations since that's already the case now. Will merge soon.

The `@contractmath` can be added later if people really want it though a `@fastmath` - `@finite_floating_point` will likely be more useful even though it's not supported by llvm afaict.
